cmd/check-markdown-rs/src/main.rs

```diff
@@ -0,0 +1,318 @@
+    );
+
+    Ok(())
+}
+
+// Recursive function to print the nodes of the AST
+fn print_nodes<'a>(
+    node: &'a Node<'a, RefCell<Ast>>,
+    arena: &'a Arena<Node<'a, RefCell<Ast>>>,
+    level: usize,
+) {
+    // Create an indent string based on the current recursion level
+    let indent = "  ".repeat(level);
+    // Match the node value to handle different node types
+    match &node.data.borrow().value {
+        // If the node is a Heading node, print its level and text
+        NodeValue::Heading(heading) => println!(
+            "{}Header (level {}): {}",
+            indent,
+            heading.level,
+            get_node_text(&node)
+        ),
+        // If the node is a Text node, print its content
+        NodeValue::Text(text) => println!(
+            "{}Text: {}",
+            indent,
+            String::from_utf8_lossy(text.as_bytes())
+        ),
+        // If the node is a Link node, print its title and URL
+        NodeValue::Link(link) => println!(
+            "{}Link: [{}]({})",
+            indent,
+            String::from_utf8_lossy(link.title.as_bytes()),
+            String::from_utf8_lossy(link.url.as_bytes())
+        ),
+        // For other node types, print the debug output
+        _ => println!("{}{:?}", indent, node.data.borrow().value),
+    }
+    // Iterate through the children of the current node and recursively print them
+    for child in node.children() {
+        print_nodes(&child, arena, level + 1);
+    }
+}
+
+// Function to get the text content of a node
+fn get_node_text<'a>(node: &'a Node<'a, RefCell<Ast>>) -> String {
+    // Iterate through the children of the node
+    node.children()
+        // Filter and map each child node to its text content, if it's a Text node
+        .filter_map(|child| match &child.data.borrow().value {
+            NodeValue::Text(text) => Some(String::from_utf8_lossy(text.as_bytes()).into_owned()),
+            _ => None,
+        })
+        // Collect the text content into a vector of strings
+        .collect::<Vec<String>>()
+        // Join the strings together to form a single string
+        .join("")
+}
+
+// Function to validate the structure of the document
+fn validate_document_structure<'a>(root: &'a Node<'a, RefCell<Ast>>) -> Result<(), String> {
+    let mut last_seen_level = 0;
+    let mut h1_count = 0;
+
+    let mut heading_nodes = Vec::new();
+    find_heading_nodes(root, &mut heading_nodes);
+
+    for node in heading_nodes {
+        match &node.data.borrow().value {
+            NodeValue::Heading(heading) => {
+                let level = heading.level;
+
+                if level == 1 {
+                    h1_count += 1;
+                    if h1_count > 1 {
+                        return Err(String::from("Error: More than one H1 heading found."));
+                    }
+                }
+
+                if level > last_seen_level + 1 {
+                    return Err(String::from("Error: Invalid heading level order."));
```

Comment:
- Can you remove the `Error: ` prefix from the errors? The type of these objects `Result::Err` already encodes that they are errors so it's best to let the higher levels handle adding prefixes like this (see comment on that topic below).
- Can you include the heading in the error being returned here? In fact, ideally, you'd also add the line number and the filename to the error (either here or at a higher level) to make it 100% clear to the user which part of which file contains the error. This is particularly important when the tool starts to jump between files when it finds a cross-file link, and then finds, for example, an invalid link in another file _with the same name as a link in the initial file_ (yep - I've been there and done that ;)

Reply:
This should be fixed now!